package/MDAnalysis/coordinates/PDB.py

```diff
@@ -851,6 +854,11 @@ def _write_pdb_bonds(self):
             con[a2].append(a1)
             con[a1].append(a2)
 
+        if not all(ag_ind <= 100000 for ag_ind in con.keys()):
```

Comment:
It ensures every entry of CONECT should be valid.

Reply:
rather than iterating through the dictionary again, you could do this check as you populate the dictionary

Reply:
Would you suggest move it even earlier? The majority of time spend here is constructing the `bondset`.
for a system with # Bond: 430447 # Atom: 579699
```python
%lprun -f mda.coordinates.PDB.PDBWriter._write_pdb_bonds u.atoms.write('test.pdb')
Total time: 131.148 s
File: /home/scottzhuang/mdanalysis/package/MDAnalysis/coordinates/PDB.py
Function: _write_pdb_bonds at line 808

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   808                                               def _write_pdb_bonds(self):
   809                                                   """Writes out all the bond records"""
   810         1          3.0      3.0      0.0          if self.bonds is None:
   811                                                       return
   812                                           
   813         3         11.0      3.7      0.0          if (not hasattr(self, "obj") or
   814         1          2.0      2.0      0.0                  not self.obj or
   815         1    3598068.0 3598068.0      2.7                  not hasattr(self.obj.universe, 'bonds')):
   816                                                       return
   817                                           
   818         1  127232750.0 127232750.0     97.0          bondset = set(itertools.chain(*(a.bonds for a in self.obj.atoms)))
   819         1          3.0      3.0      0.0          if self._reindex:
   820         1          1.0      1.0      0.0              index_attribute = 'index'
   821         2     308498.0 154249.0      0.2              mapping = {
   822                                                           index: i
   823         1         21.0     21.0      0.0                  for i, index in enumerate(self.obj.atoms.indices, start=1)
   824                                                       }
   825         1       5797.0   5797.0      0.0              atoms = np.sort(self.obj.atoms.indices)
   826                                                   else:
   827                                                       index_attribute = 'id'
   828                                                       mapping = {id_: id_ for id_ in self.obj.atoms.ids}
   829                                                       atoms = np.sort(self.obj.atoms.ids)
   830         1          3.0      3.0      0.0          if self.bonds == "conect":
   831                                                       # Write out only the bonds that were defined in CONECT records
   832         2          6.0      3.0      0.0              bonds = (
   833                                                           (
   834                                                               getattr(bond[0], index_attribute),
   835                                                               getattr(bond[1], index_attribute),
   836                                                           )
   837         1          1.0      1.0      0.0                  for bond in bondset if not bond.is_guessed
   838                                                       )
   839                                                   elif self.bonds == "all":
   840                                                       bonds = (
   841                                                           (
   842                                                               getattr(bond[0], index_attribute),
   843                                                               getattr(bond[1], index_attribute),
   844                                                           )
   845                                                           for bond in bondset
   846                                                       )
   847                                                   else:
   848                                                       raise ValueError("bonds has to be either None, 'conect' or 'all'")
   849                                           
   850         1          5.0      5.0      0.0          con = collections.defaultdict(list)
   851         2       2718.0   1359.0      0.0          for a1, a2 in bonds:
   852         2          5.0      2.5      0.0              if not (a1 in mapping and a2 in mapping):
   853                                                           continue
   854         2          4.0      2.0      0.0              if mapping[a1] >= 100000 or mapping[a2] >= 100000:
   855         1        130.0    130.0      0.0                  warnings.warn("Atom with index >=100000 cannot write "
   856                                                                         "bonds to PDB CONECT records.")
   857         1          1.0      1.0      0.0                  return
   858         1          3.0      3.0      0.0              con[a2].append(a1)
   859         1          1.0      1.0      0.0              con[a1].append(a2)
   860                                           
   861                                                   conect = ([mapping[a]] + sorted([mapping[at] for at in con[a]])
   862                                                             for a in atoms if a in con)
   863                                           
   864                                                   for c in conect:
   865                                                       self.CONECT(c)
```

Reply:
Ouch yeah that's 2 minutes making that set.  @hmacdope an example of where bond performance is terrible currently.

Reply:
Roger roger. Good to have as a bench/test case.

Reply:
@hmacdope  If you are interested. https://github.com/yuxuanzhuang/MDA_benchmark_sys